### ISSUE TITLE
Support OWTF on Windows

### ISSUE THREAD

**Issue by anantshri** [owtf/owtf#448]
Consider this a meta issue opened to start gradual support of OWTF in windows.

starting we have following two major problems

https://github.com/owtf/owtf/tree/master/dictionaries/fuzzdb/fuzzdb-1.09/docs/misc/unicode-security contains a file and folder which contains ":" which makes it not creatable in windows.

simmilarly we use "aux" as a type of plugin which is a protected name in windows.

I wanted to work on this and hence started with my own fork of owtf however seems like i am not getting time and this project keep getting more and more commits in it. so can i ask someone to please have a look at these two issues, Once that is done we would be able to atleast clone owtf in windows and then we can start with testing for what is working and what not and the slowly look at supporting windows also.

**Comment by 7a** [comment]
+1

**Comment by anantshri** [comment]
Just asking do we need this folder https://github.com/owtf/owtf/tree/master/dictionaries/fuzzdb/fuzzdb-1.09/docs/misc/unicode-security can we remove this folder all together from the core. That should help out a lot.

This will leave with issues like net and aux not creating coz they are reserved names in windows.

**Comment by delta24** [comment]
@anantshri Ok, so we can remove that folder. Can we change the naming scheme for net and aux? Is it feasible?

**Comment by anantshri** [comment]
Last time when i tried renaming the folder and all references inside somehow owtf was not working. So i suppose its feasable but it will just take some time.

as per discussion with @7a we decided to keep net as network and aux as auxillary.

**Comment by delta24** [comment]
@anantshri Do you have to time to tackle this issue? I think you are best placed to do it :).

**Comment by anantshri** [comment]
Like i keep saying i can give it a shot but not sure of a deadline or timeline when i will get it done.

**Comment by delta24** [comment]
@anantshri  Sure no problem :)

**Comment by delta24** [comment]
List of things to be done:

- [x]  Initial git clone problems - fuzzdb, aux->auxillary, net->network
- [ ] Research how to package the whole installer - NSI installer ? or simple installation like Kali, Debian?
- [ ] Powerscripts support

**Comment by sachinkamath** [comment]
@delta24 I'd like to take this up. I am thinking of more like a simple installer (maybe a bat?). But the current problem is, all the installer scripts in the repo are written for Linux . Things like /dev/null and stuff will have to be re-written. 
The best thing to do'd be a separate branch for Windows. And why is this out of the OWTF Quality Release? Time issues?

**Comment by sachinkamath** [comment]
## List of Concerns + Bugs - OWTF on Windows

Here are a list of problems I found when trying to install OWTF on Windows.

* Current installation script ( install.py ) is for Linux only. We will have to add support for Windows by either tweaking it or making an entirely new one for Windows, exclusively.

### PIP installation issues

* **[CRITICAL]** pexpect==3.3 does not support Windows atm.

![2016-04-05 19_53_57-administrator_ windows powershell x86](https://cloud.githubusercontent.com/assets/9546091/14285915/de282ee6-fb6a-11e5-9b8b-85a073b385b7.png)

 Upon further investigation, realized that the version is the problem. Windows support was introduced in later versions. It'd be best to upgrade pexpect to 4.0.1 or use [Wpexpect](https://gist.github.com/anthonyeden/8488763) if upgrade breaks OWTF. ( *Requires investigation* )

* pycurl==7.19.5 installation fails on Windows.

![2016-04-05 20_30_17-select administrator_ windows powershell x86](https://cloud.githubusercontent.com/assets/9546091/14286639/5fe8d12c-fb6d-11e5-9f24-27c9fa3a59c7.png)

Happens because curl doesn't come inbuilt with Windows. Installing curl binaries solves the problem. Interestingly, pycurl version 7.43.0 did not throw an error even when curl package was not present. 


**(Will be updated as I take futher steps)**

**Comment by sachinkamath** [comment]
## List of Concerns + Bugs - OWTF on Windows

Here are a list of problems I found when trying to install OWTF on Windows.

* Current installation script ( install.py ) is for Linux only. We will have to add support for Windows by either tweaking it or making an entirely new one for Windows, exclusively.

### PIP installation issues

* **[CRITICAL]** pexpect==3.3 does not support Windows atm.

![2016-04-05 19_53_57-administrator_ windows powershell x86](https://cloud.githubusercontent.com/assets/9546091/14285915/de282ee6-fb6a-11e5-9b8b-85a073b385b7.png)

 Upon further investigation, realized that the version is the problem. Windows support was introduced in later versions. It'd be best to upgrade pexpect to 4.0.1 or use [Wpexpect](https://gist.github.com/anthonyeden/8488763) if upgrade breaks OWTF. ( *Requires investigation* )

* pycurl==7.19.5 installation fails on Windows.

![2016-04-05 20_30_17-select administrator_ windows powershell x86](https://cloud.githubusercontent.com/assets/9546091/14286639/5fe8d12c-fb6d-11e5-9f24-27c9fa3a59c7.png)

Happens because curl doesn't come inbuilt with Windows. Installing curl binaries solves the problem. Interestingly, pycurl version 7.43.0 did not throw an error even when curl package was not present. 


* No support for wget in Windows

Wget commands fail in Windows as it doesn't come with wget. I tried downloading [Windows Wget](http://gnuwin32.sourceforge.net/packages/wget.htm) , but the python `os.system` was not respecting the PATH variables and was reporting `wget` as not present even-though I had correctly installed it. Copy pasting the exact command into the Command Prompt worked. We should consider moving to [python-wget](https://pypi.python.org/pypi/wget) as it is more python-ic and reduces the dependency on pre-installed packages. (*This can be a new issue*)


**(Will be updated as I take further steps)**

**Comment by delta24** [comment]
@anantshri @7a @marioskourtesis @tunnelshade After Appsec EU discussions, I am more inclined to shift OWTF core to Docker as Marios suggested because porting to Windows will mean more code to maintain, in addition to multiple `python` versions (2.7/3.x).

What do you guys think? We can have an OWTF container, Postgresql containers, and a OWTF proxy container (as to decouple components).

**Comment by anantshri** [comment]
+1 for docker options. However one of the other major reason for opening this ticket was that at the time of opening the ticket owtf was not even git clonable on windows machines. I would suggest we go ahead with docker approach but we atleast make sure its clonable at all OS.

**Comment by delta24** [comment]
Now that the Docker support is complete, I am closing this now - I'll be updating the Docker repository in a few days.
